src/ui/carousel/carousel.ts

```diff
@@ -0,0 +1,548 @@
+  private readonly raf: Raf;
+  private readonly eventManager: EventManager;
+  private transitionTarget: HTMLElement;
+  private lastActiveSlide: HTMLElement;
+  private autoplayTimeout: number;
+  private syncedCarousels: Set<Carousel>;
+
+  /**
+   * @param container Parent element of slides.
+   * Element that acts as the carousel, in which all slides are contained.
+   * Does not need to be the direct parent.
+   *
+   * @param slides HTMLElements containing slide content.
+   *
+   * @param condition Under what conditions the carousel should run
+   * @param activeCssClass Class to apply to active slide.
+   * @param beforeCssClass Class to apply to slides before active slide.
+   * @param afterCssClass Class to apply to slides after active slide.
+   * @param distanceToActiveSlideAttr Slide attribute on which the index
+   *     distance to the currently active slide is set.
+   * @param allowLooping Whether the carousel should be allowed to loop.
+   * @param transition How the Carousel should transition.
+   * @param autoplaySpeed How long the carousel should stay on a slide (in ms)
+   *     before transitioning to the next slide.
+   */
+  constructor(
+      container: HTMLElement,
+      slides: HTMLElement[],
+      {
+        condition = () => true,
+        activeCssClass = DefaultCssClass.ACTIVE_SLIDE,
+        beforeCssClass = DefaultCssClass.BEFORE_SLIDE,
+        afterCssClass = DefaultCssClass.AFTER_SLIDE,
+        distanceToActiveSlideAttr = DEFAULT_DISTANCE_TO_ACTIVE_SLIDE_ATTR,
+        allowLooping = true,
+        transition = null,
+        autoplaySpeed = null
+      }: CarouselOptions = {}
+  ) {
+    if (slides.length < 1) {
+      throw new Error('Cannot start carousel without slides');
+    }
+    this.raf = new Raf(() => this.onRaf());
+    this.activeCssClass = activeCssClass;
+    this.beforeCssClass = beforeCssClass;
+    this.afterCssClass = afterCssClass;
+    this.distanceToActiveSlideAttr = distanceToActiveSlideAttr;
+    this.loop = allowLooping;
+    this.condition = condition;
+    this.container = container;
+    this.lastActiveSlide = null;
+    this.eventManager = new EventManager();
+    this.slides = slides;
+    if (typeof transition === 'string') {
+      switch (transition) {
+        case 'css':
+          this.transition = new CssClassesOnly();
+          break;
+        case 'draggable':
+          this.transition = new DraggableSlide();
+          break;
+        default:
+          throw new Error(
+              `Unrecognized transition type "${transition}" passed to ` +
+              `Carousel.`);
+      }
+    } else {
+      this.transition = transition || new CssClassesOnly();
+    }
+    this.transitionTarget = null;
+    this.autoplaySpeed = autoplaySpeed;
+    this.autoplayTimeout = null;
+    this.syncedCarousels = new Set<Carousel>([this]);
+
+    this.init();
+  }
+
+  /**
+   * Returns the last known active slide.
+   */
+  getLastActiveSlide(): HTMLElement {
```

Comment:
What's the use case for this?

Reply:
DraggableSlide calls it as a tie breaker when determining the active slide resulting from a drag.

If two slides are tied for distance as closest to the center, it will default to the one that was last active, not crucial, can remove.